compiler/ir/backend.jvm/lower/src/org/jetbrains/kotlin/backend/jvm/lower/JvmInlineClassLowering.kt

```diff
@@ -62,16 +71,292 @@ private class JvmInlineClassLowering(context: JvmBackendContext) : JvmValueClass
+            return super.visitTypeOperator(expression)
+        }
+
+        if (expression.typeOperand.isInlineChildOfSealedInlineClass()) {
+            val transformed = super.visitTypeOperator(expression) as IrTypeOperatorCall
+            val top = transformed.typeOperand.findTopSealedInlineSuperClass()
+            val isCheck = context.inlineClassReplacements.getIsSealedInlineChildFunction(top to transformed.typeOperand.classOrNull!!.owner)
+            val underlyingType = transformed.typeOperand.unboxInlineClass()
+            val currentScopeSymbol = (currentScope?.irElement as? IrSymbolOwner)?.symbol
+                ?: error("${currentScope?.irElement?.render()} is not valid symbol owner")
+
+            when (transformed.operator) {
+                IrTypeOperator.INSTANCEOF -> {
+                    // if (top != null) false else Top.is-Child(top)
+                    with(context.createIrBuilder(currentScopeSymbol)) {
+                        return irBlock {
+                            val tmp = irTemporary(transformed.argument)
+                            +irIfNull(
+                                context.irBuiltIns.booleanType, irGet(tmp), irFalse(),
+                                irCall(isCheck.symbol).also {
+                                    it.putValueArgument(
+                                        0,
+                                        coerceInlineClasses(irGet(tmp), top.defaultType, context.irBuiltIns.anyNType)
+                                    )
+                                }
+                            )
+                        }
+                    }
+                }
+                IrTypeOperator.NOT_INSTANCEOF -> {
+                    // if (top != null) true else Top.is-Child(top).not()
+                    with(context.createIrBuilder(currentScopeSymbol)) {
+                        return irBlock {
+                            val tmp = irTemporary(transformed.argument)
+                            +irIfNull(
+                                context.irBuiltIns.booleanType, irGet(tmp), irTrue(),
+                                irNot(
+                                    irCall(isCheck.symbol).also {
+                                        it.putValueArgument(
+                                            0,
+                                            coerceInlineClasses(irGet(tmp), top.defaultType, context.irBuiltIns.anyNType)
+                                        )
+                                    }
+                                )
+                            )
+                        }
+                    }
+                }
+                IrTypeOperator.CAST -> {
+                    // if (Top.is-Child(top)) top else CCE
+                    return generateAsCheck(currentScopeSymbol, transformed, top, isCheck, underlyingType) {
+                        irCall(this@JvmInlineClassLowering.context.ir.symbols.throwTypeCastException).also {
```

Comment:
Let's use `java.lang.ClassCastException` instead. We can use a more specific exception type later if needed. `kotlin.TypeCastException` is now only used in case of API version 1.3 ([KT-22275](https://youtrack.jetbrains.com/issue/KT-22275/Unify-exceptions-from-null-checks))